### ISSUE TITLE
Problems with tables and external functions in MSL

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#76]
**Reported by kurzbach on 8 May 2008 04:51 UTC**
In Modelica.Blocks.Tables there are table implementations which call external functions. There are several problems with them:
1. There is no free C code for them.
2. The interface is not described.
3. There is no function to cleanup the dynamical allocated memory.

Proposal:
1. There needs to be a tool independent free implementation (atleast to show the principles, with reduced error handling) which is allowed to be modified to add tool specific things. 
2. The interface and functionality of the functions have to be described clearly.
3. The implementation should make use of external objects to have clearly defined constructors and destructors.

In the future these functions could be implemented solely in Modelica, but this can be done later.

The descibed problems apply analogously for similar places where external functions are used in the MSL. 

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/76

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 8 May 2008 07:18 UTC**
The interface of the external tables is described in the provided header file:
[\Modelica\C-Sources\ModelicaTables.h](https://svn.modelica.org/Modelica/trunk/Modelica/C-Sources/ModelicaTables.h)

A function to free the allocated memory is already defined in ModelicaTables.h, however it is not called in the table functions (so this is a bug). It is most likely not safe to call the de-allocation function in "when terminal(..) then ..", since it is not guaranteed that sorting is performed in such a way that no table function is called afterwards. Therefore, the table implementation should be rewritten as ExternalObject. Due to timing reasons, this was not made for Modelica 3.0, but should now be made for Modelica 3.1.

I agree, that it is important to have a tool independent free implementation. The question is who is performing the work and in the past there have been no volunteers (the problem is that it is quite a lot of work to implement smooth tables in C). My favorite approach is to write nearly everything in Modelica, because implementation, support and further development is much less work. There is also already a prototype availabe under https://svn.modelica.org/Modelica/trunk/ModelicaDesign/Modelica_Interpolation. The trouble is that with this prototype the table arrays are held as Modelica arrays and this is both unefficient for large tables and the table size is fixed after translation. Since about 2 years, there is a proposal to fix this without a change in Modelica by having support of automatic caching in Modelica tools: Assume that an interpolation y=f(x) shall be implemented with B-Splines, the basic idea is then

```
y = interpolate(makeSpline(readFromFile("table.mat")), x)
```
Without caching, this statement means that whenever interpolation takes place, the table is read from file "table.mat" and it is transformed to a spline (so the spline coefficients are computed) and then interpolation takes place. This is, of course, extremely unefficient. However, a tool could recognize that the arguments to "readFromFile(..)" and to "makeSpline" are constants or parameters and could therefore call these functions only once, automatically. Furthermore, since the result arrays of these functions are not accessed anywhere, a tool could just allocate storage for them (even during initialization, and not during translation) without any overhead. 

If at least one tool vendor provides this feature so that testing can be performed, DLR will finalize the free implementation of Modelica_Interpolation, so that the current functionality of the table functions are covered. For a pure C implementation, DLR has currently not enough resources (it might be that others could offer an implementation). Note, this approach has a lot of advantages and it is very easy to provide much more functionality to tables as is currently provided. The "killer issue" is to have appropriate support of automatic caching in the tools.

To my knowledge, there is only one other place in MSL3.0, where no free implementation is offered: Modelica.Mechanics.MultiBody.Visualizers.Advanced.Shape to visualize 3-dim. objects. DLR is currently working on a very good alternative that is tool independent (and is much better as the current Shape object). It might be that a proposal can be presented at the next design meeting.

If there are other places in MSL3.0 where external functions are called without a free, available implementation, please list them in order that it can be fixed.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 21 Jan 2009 10:00 UTC**
 Martin Otter wrote::
 Not possible to fix quickly (large amount of work). I started to experiment with a good solution, but I got stuck by higher priority issues. Will propose a solution in mid of February. Of course, anyone is free to provide/propose other solutions.

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 21 Jun 2009 10:09 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 25 Nov 2009 11:53 UTC**
Milestone MSL3.1.1 deleted

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 Jul 2010 16:23 UTC**
There is not enough time to fix this in MSL3.2. Needs to be fixed in the follow up version.

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 6 Sep 2012 11:47 UTC**

**Comment by modelica-trac-importer** [comment]
**Changelog removed by dietmarw on 6 Sep 2012 11:47 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 17 Dec 2012 13:25 UTC**
Duplicate of #129 and will be fixed there.

**Comment by modelica-trac-importer** [comment]
**Modified by kurzbach on 20 Dec 2016 13:30 UTC**
